### ISSUE TITLE
No phasing blocks

### ISSUE THREAD

**Issue by wsstoregene** [whatshap/whatshap#502]
Hello,

I have been trying to phase single WGS genome files but no variants were used and consequently no blocks were phased.

I am using block genome VCFs. Not sure what the problem could be. Could  you please point me to any direction?
Please, have a look at the screenshot I have attached to this ticket.

Thank you.

![Screenshot 2023-11-15 at 12 21 20 (1)](https://github.com/whatshap/whatshap/assets/136618049/3d6e6841-31ad-422a-9fa3-cc8b7ea36b1e)

**Comment by schrins** [comment]
From first look I would assume that the variants in your VCF are not genotyped as heterozygous and thus all skipped. Can you check whether the GT field in the VCF contains `0/1`-entries?

**Comment by wsstoregene** [comment]
Hi,
I have run one sample including chr1 only. 

This is the output:

command line: whatshap phase -o  ASD_PHASED_WHATSHAP.vcf --reference=/mnt/sda/bin/ucsc-hg19-fasta/ucsc.hg19.fasta ASDE.unblocked.vcf.gz ASDE.md.cram         

                                                                           
This is WhatsHap 2.0 running under Python 3.10.12
Working on 1 sample from 1 family

# Working on contig chr1 in individual 217_ASDE
Found 0 usable heterozygous variants (0 skipped due to missing genotypes)
Found 0 reads covering 0 variants
Kept 0 reads that cover at least two variants each
Selected 0 most phase-informative reads covering 0 variants
Best-case phasing would result in 0 non-singleton phased blocks (0 singletons). 
Phasing 1 sample by solving the MEC problem ... 
No. of phased blocks: 0

# Resource usage
Maximum memory usage: 1.331 GB
Time spent reading BAM/CRAM:                  222.5 s
Time spent parsing VCF:                         8.4 s
Time spent selecting reads:                     0.0 s
Time spent phasing:                             0.0 s
Time spent writing VCF:                         8.6 s
Time spent finding components:                  0.0 s
Time spent on rest:                            14.8 s
Total elapsed time:                           254.4 s


I have also checked. The output is a gVCF produced by haplotypecaller:
- zgrep "0/1" ASDE.unblocked.vcf.gz | wc -l = 300
- zgrep "0|1" ASDE.unblocked.vcf.gz | wc -l = 95

**Comment by schrins** [comment]
This sounds strange. Since you use version 2.1, it should by default consider indels, which were ignored before 2.0. What I could still think of are alleles that whatshap cannot handle. For diploid phasing, multi-allelic variants are discarded, because the diploid algorithm cannot handle them. Also, if the REF and ALT sequences are not made of nucleotide bases, (e.g. `.` or `N`), I think these variants are also discarded. Could you rerun the data but with a `--debug` flag behind `whatshap`? So `whatshap --debug phase ...`

Maybe this gives us more hints about where the variants are lost.

**Comment by wsstoregene** [comment]
Hi,
This is the output in verbose (debug) mode:


whatshap --debug phase  -o  ASDE_PHASED_WHATSHAP.vcf --reference=/mnt/sda/bin/ucsc-hg19-fasta/ucsc.hg19.fasta ASDE.haplotypecaller.vcf.gz  ASDE.md.cram
This is WhatsHap 2.0 running under Python 3.10.12
DEBUG: Creating PhasedInputReader
DEBUG: Read groups in CRAM/BAM header: [{'ID': 'H7HLGDSXY.ASDE1234.1', 'PU': '1', 'SM': '217_ASDE', 'LB': 'ASDE', 'DS': '/data/pipeline-data/fixed-input/hg19.fa', 'PL': 'ILLUMINA'}]
DEBUG: Creating PhasedVcfWriter
DEBUG: Reading the input VCF to find possibly missing headers
DEBUG: Missing contigs: []
DEBUG: Missing formats: ['AD']
DEBUG: Missing infos: []
DEBUG: Found 1 sample(s) in the VCF file.
Working on 1 sample from 1 family
DEBUG: Parsed 0 SNVs and 48721 non-SNVs. Also found 548 multi-ALTs.

# Working on contig chr1 in individual 217_ASDE
DEBUG: Using maximum coverage per sample of 15X
Found 0 usable heterozygous variants (0 skipped due to missing genotypes)
DEBUG: Reading alignments for sample '217_ASDE' on chromosome chr1 and detecting alleles ...
Found 0 reads covering 0 variants
Kept 0 reads that cover at least two variants each
DEBUG: Reducing coverage to at most 15X by selecting most informative reads ...
DEBUG: Running read selection for 0 reads covering 0 variants (bridging ON)
Selected 0 most phase-informative reads covering 0 variants
Best-case phasing would result in 0 non-singleton phased blocks (0 singletons). 
DEBUG: ... would be 0 non-singleton phased blocks without read selection
DEBUG: Variants covered by at least one phase-informative read in at least one individual after read selection: 0
Phasing 1 sample by solving the MEC problem ...
DEBUG: MEC cost: 0
DEBUG: Finding connected components ...
No. of phased blocks: 0
DEBUG: Writing phasing result to output VCF
DEBUG: Chromosome 'chr1' finished

# Resource usage
Maximum memory usage: 0.583 GB
Time spent reading BAM/CRAM:                  219.6 s
Time spent parsing VCF:                         0.5 s
Time spent selecting reads:                     0.0 s
Time spent phasing:                             0.0 s
Time spent writing VCF:                         0.5 s
Time spent finding components:                  0.0 s
Time spent on rest:                             1.6 s
Total elapsed time:                           222.3 s

**Comment by wsstoregene** [comment]
Following my previous post, this is a visualisation of the genome vcf format derived from haplotypecaller using the -ERC GVCF function. The pipeline I am currently using is the nf-core/sarek pipeline.

I have tried to modify the info present in the vcf but it did not help.

########################## genome VCF
<img width="1610" alt="Screenshot 2023-12-13 at 10 08 27" src="https://github.com/whatshap/whatshap/assets/136618049/c15f126c-042c-4f1f-a5a4-2515b2955e41">
<img width="1611" alt="Screenshot 2023-12-13 at 10 09 06" src="https://github.com/whatshap/whatshap/assets/136618049/6f02f0ee-ae50-4508-b0dc-2ccd652aff6e">

**Comment by marcelm** [comment]
Hi @wsstoregene, thanks for reporting. Just as a note, please avoid pasting screenshots of text as they are not as accessible. This is also written in the instructions when you open a new issue.

`whatshap phase` cannot work with this type of gVCF file at the moment. We fixed gVCF handling some time ago in #289, but it seems that newer gVCFs contain `<NON_REF>` in the ALT field for all positions, which poses a probem: When the VCF parser sees something like `GC,<NON_REF>`, it considers this to be a multiallelic variant (since there are two "alleles" `GC` and `<NON_REF>`). Multiple alternative alleles aren’t supported by `whatshap phase`, so the variant is simply skipped. The end result of course is that *all* relevant variants are skipped.

I am not sure when one of us has time to fix this.
